### DISCUSSION TITLE
Targets for Notebooks

### DISCUSSION THREAD

**Discussion by nviets** [ropensci/targets#469] [Ideas]
Are there any plans for recommended strategies to integrate targets with Rmd Notebooks? Python has [dataflow](https://dataflownb.github.io/), and Julia has [Pluto](https://github.com/fonsp/Pluto.jl). Targets seems like a natural starting point for similar functionality in Rmd.

**Comment by wlandau** [comment]
Through [`tarchetypes`](https://docs.ropensci.org/tarchetypes/), `targets` seamlessly integrates with R Markdown, including parameterized reports. This setup allows you to run reports as part of a pipeline, knit them interactively, and easily go back and forth between these two options. In this scenario, a good R Markdown report is 99% prose and 1% R code, taking advantage of targets you already computed upstream in the pipeline. Details:

* https://books.ropensci.org/targets/files.html#literate-programming
* https://books.ropensci.org/targets/files.html#parameterized-r-markdown
* https://docs.ropensci.org/tarchetypes/reference/index.html#section-literate-programming-targets
* https://docs.ropensci.org/tarchetypes/reference/index.html#section-literate-programming-utilities

**Reply by nviets** [reply]
Thanks for the reply! We love targets' integration with Markdown in the sense of generating parametrized reports from the targets cache. In the above, I was thinking of targets more as a backend to power an interactive document experience in the sense of Pluto and dataflow. A lot of users program interactively Rmd notebooks, and there isn't really a solution today in Rmd similar to Pluto and Dataflow. If we could turn rmd chunks into the targets and have chunks activate a tar_make command, we'd get all the benefits of targets paired with the interactive experience of notebooks.

**Reply by wlandau** [reply]
> If we could turn rmd chunks into the targets...

I usually push back against R Markdown-driven development because literate programming is not designed for computationally intense work. That said, at the interface level, it is possible to represent a pipeline as an R Markdown file.

<pre><code>
---
title: "Example report"
output: tarchetypes::tar_render_pipeline
---

```{r, memory = "transient"}
library(tidyverse)
get_data <- function() {
  # ...
}
run_analysis <- function(data) {
  # ...
}
```

Prose for the first target:

```{r analysis_target, format = "qs"}
data <- get_data()
run_analysis(data)
```

Prose for the second target:

```{r summary_target, format = "parquet"}
analysis_target %>%
  group_by(model, method) %>%
  summarize(accuracy = hits / total) %>%
  ungroup()
```
</code></pre>

A preprocessor, maybe a new function in `tarchetypes`, could convert this R Markdown file into a `_targets.R` file. 

```r
# _targets.R
targets::tar_option_set(memory = "transient") # from the options of the unnamed chunks

# Unnamed chunks:

library(targets)
library(tidyverse)

get_data <- function() {
  # ...
}
run_analysis <- function(data) {
  # ...
}

# Named chunks are targets:
list(
  tar_target(
    analysis_target, {
      data <- get_data()
      run_analysis(data)
    },
    format = "qs" # from the chunk options
  ),
  tar_target(
    analysis_target, {
      analysis_target %>%
        group_by(model, method) %>%
        summarize(accuracy = hits / total) %>%
        ungroup()
    },
    format = "parquet" # from the chunk options
  )
)
```

Then all the functions in `targets` would be available, from `tar_make()` to `tar_visnetwork()`. And maybe it might be possible to configure the `knitr` engine to run the preprocessor and then `tar_make()`.

Is that pretty much what you had in mind?

**Reply by wlandau** [reply]
Edit: IIRC, the `output` field of the YAML front matter lets us take control of how `knitr` runs the report. I added a placeholder above for `output: tarchetypes::tar_render_pipeline`.

**Reply by wlandau** [reply]
Running an individual code chunk would be tricky though. At the very least, it could be shorthand for `tar_make(names = "target_of_chunk")`, which would run/check the selected target after running/checking the previous ones. I wonder if `output: tarchetypes::tar_render_pipeline` in the YAML opens up a way to implement that.

**Reply by wlandau** [reply]
Target factories would need to be handled a bit differently, maybe with a chunk option, e.g. `{r, ..., factory = stantargets::tar_stan_mcmc}`

**Reply by wlandau** [reply]
The YAML front matter could also let the user choose `tar_make()`, `tar_make_clustermq()`, or `tar_make_future()`, and supply arguments to the `tar_make*()` function of choice. This makes me think the custom `knitr` engine should both preprocess the R Markdown and run the pipeline, especially if we have an alternative preprocessing-only function exported from `tarchetypes`.

**Reply by nviets** [reply]
Thanks @wlandau ! Your outline is much closer to what I had in mind. Regarding the runs of individual code chunks, have you looked at [custom chunk engines](https://bookdown.org/yihui/rmarkdown-cookbook/custom-engine.html)? Could they be used to trigger the _targets.R creation and tar_make()?

**Reply by wlandau** [reply]
Sounds like custom chunk engines are what I was looking for. Do they operate independently of the `output` engine declared in the YAML front matter? Is it okay to set the chunk engine from within the `output` engine, or is it an anti-pattern?

**Reply by nviets** [reply]
I believe the custom language engines are independent of the output engine. The built-in language engines are [here](https://github.com/yihui/knitr/blob/master/R/engine.R) as an example of how a non-R chunk is run within the context of the overall Rmd.

**Reply by cderv** [reply]
@wlandau I try answer your question about custom knitr engine if you want. 
I don't know if you went this way or not. 

Re: your comment above

* Custom engine are independent of output format you set in YAML. Custom engine are only a knitr think and define how to execute the code, and how to output the result. 
* Not sure what setting chunk engine in the output engine means but basically a custom engine function would be define in a package and register to knitr as a engine either when package is loaded or using a helper function to set in the setup chunk. 

Please do not hesitate to reach out on this if I can help.

**Reply by wlandau** [reply]
Thanks, @cderv! I really appreciate being able to send questions your way. 

Here is what I am thinking at the moment:

1. A custom chunk engine to emulate the pure function behavior of `targets`. If a chunk has `type = "target"` as an option, run it inside `local()` and assign the return value to a new variable whose name is the chunk name. Would a custom chunk engine be capable of this?
2. A `tarchetypes` function that crawls through one or more Rmd files and translates the code chunks into `_targets.R` etc. for a `targets` pipeline. This requires the ability to purl an Rmd in a way that separately identifies the code and options of each chunk. Does `rmarkdown` or `knitr` expose a way to do this?
3. Somehow incorporate the user's original prose in rendered Rmd's that can print out the return values of targets. This last part is less clear to me, but maybe we could programmatically replace code chunks with `tar_read()` and render them as `tar_render()` outputs from the pipeline.

**Reply by wlandau** [reply]
So there are multiple types of functionality we could offer users:

1. Prototyping: emulate the pure-function-like behavior of a target when the user runs a chunk interactively in RStudio.
2. Construction: write target code and function scripts to build up a valid `_targets.R` file.
3. Execution: actually run the pipeline, e.g. `tar_make()`.
4. Reporting: render an output document that prints the return values of built target next to sections of the user's prose (i.e. true literate programming).

(1) and (2) would be the biggest wins because pipeline construction is still unfamiliar/confusing/mysterious for many new users. (3) is already easy because of `tar_make()`/ `tar_visnetwork()`/`tar_read()`, and (4) is already covered by https://books.ropensci.org/targets/files.html#literate-programming. And (1) and (2) can be solved using custom chunk engines, which look easy to implement on my end.

**Reply by wlandau** [reply]
# Sketch of the Target Markdown system

## Target engine

Following https://bookdown.org/yihui/rmarkdown-cookbook/custom-engine.html, a target engine should solve most of (1) and (2) above:

1. If invoked interactively for prototyping, run the code inside `local()` and assign the return value to a variable with the target name. Assign to the environment currently registered with the R Markdown document, if such a thing is explicit (I need to check).
2. If invoked non-interactively for pipeline construction, e.g. the Knit button in RStudio, then put the code in a `tar_target()` call and write that to a script file.

For example, the user would write something like this in the R Markdown document:

<pre><code>
```{tar_target data, format = "feather"}
raw <- download_data_data()
raw %>%
  remove_missing() %>%
  dplyr::arrange(ID)
```
</code></pre>

If you open the R Markdown in the RStudio IDE and click the green arrow to run the chunk, the code should run in `local()` and then optionally print a message telling the user what happened. After that's done, there should be a new variable `data` in your environment and no variable called `raw`. (This emulates `targets`, which prevents in-memory side effects.)
functions
If you click "Knit" in RStudio, this code chunk should write a file called `_targets_r/targets/data.R` and print a message to the rendered HTML explaining what happened:

```r
# _tar_markdown/targets/data.R
tar_target(
  name = data,
  command = {
    raw <- download_data_data()
    raw %>%
      remove_missing() %>%
      dplyr::arrange(ID)
  },
  format = "feather"
)
```

And it should overwrite `_targets.R` with this:

```r
# Generated by Tar Markdown. Do not edit by hand.
library(targets)
# Load packages, options and functions:
lapply(list.files("_tar_markdown/globals", pattern = "\\.R$", full.names = TRUE), source)
# Populate and return the target list:
lapply(list.files("_tar_markdown/targets", pattern = "\\.R$", full.names = TRUE, function(x) source(x)$value)
```

## Globals engine

The globals engine is to load packages, set options (e.g. `tar_option_set()`), and define global functions called from the targets. It should behave similarly to the target engine. Interactively, it should run just like the default R chunk engine. When you knit the whole document, it should write a script to `_targets_r/globals/your_chunk_name.R` which gets sourced in `_targets.R`.

<pre><code>
```{tar_globals your_chunk_name}
tar_option_set(packages = "tidyvrese", resources = list(bucket = "my_aws_s3_bucket_name"))

download_raw_data <- function(API_KEY = Sys.getenv("KEY_ENV"), ...) {
  # do stuff...
}
```
</code></pre>

## Target factories? 

It would be awesome if users of `stantargets` etc. could take advantage of this system. ~~Maybe the target engine should have a `factory` option to choose a target factory, with the understanding that the code inside the chunk still corresponds to a command. In the case of `stantargets::tar_stan_mcmc()`, the code in the chunk should be code to generate a dataset.~~

<pre><code>
```{tar_target my_stan_model, factory = stantargets::tar_stan_mcmc, chains = 4, iter_sampling = 4e3}
# This is some code to generate the Stan dataset,
# and it will sit in the `data` argument of `tar_stan_mcmc()`.
list(
  n = 100,
  y = rnorm(100),
  x1 = rnorm(100),
  x = splines::ns(x1)
)
```
</code></pre>

**EDIT**: on second thought, it would probably be helpful to have different engines for specific target factories. If the names align, that would help users make the transition to and from Target Markdown. So for `stantargets`, you might have something like this:

<pre><code>
```{tar_stan_mcmc my_model, chains = 4}
generate_stan_data()
```
</code></pre>

So we probably need a `knitr` engine factory that makes it easy for [R Targetopia](https://wlandau.github.io/targetopia) packages like `stantargets` to define their own target engines.

**Reply by wlandau** [reply]
# Where to implement

It would be really slick if you could just call `library(targets)` to load the chunk engines. That makes me think `targets` itself is the right home for them, as opposed to `tarchetypes`. The engines should be incredibly quick, short, and easy to implement, so it shouldn't be a big maintenance burden. It just depends on the scope creep we expect from this, `tarchetypes` can absorb tech debt more easily if things get out of hand.

**Reply by wlandau** [reply]
# Where to write the scirpts

The target chunks will write `_targets.R` at the project root as required by `targets`, and I am thinking that the scripts should be written to a new `_tar_markdown/` folder. The locations of both `_targets.R` and `_tar_markdown` should respect whatever the user chooses for `knitr::opts_knit$set(root.dir = ...)`.

**Reply by nviets** [reply]
Looks awesome @wlandau ! Will running the target chunks in the Rmd also call the `tar_make` command (or `tar_make_clustermq`)?

**Reply by wlandau** [reply]
Not in the first rollout at least, but you will be able to run the pipeline in a regular R code chunk later in the report. Sketch:

<pre><code>
```{r setup}
library(targets) # will load the chunk engines
```

```{tar_globals script_name}
download_data <- function() {...}
remove_missing <- function(raw) {...}
analyze_data <- function(data) {...}
```

```{tar_target data, format = "feather"}
raw <- download_data_data()
raw %>%
  remove_missing() %>%
  dplyr::arrange(ID)
```

```{tar_target analysis, format = "qs"}
analyze_data(data)
```

```{r}
tar_make_clustermq(workers = 4)
```
</code></pre>

**Reply by nviets** [reply]
Makes sense. Thank you!

**Reply by cderv** [reply]
Just trying to answer some questions: 

> A custom chunk engine to emulate the pure function behavior of targets. If a chunk has type = "target" as an option, run it inside local() and assign the return value to a new variable whose name is the chunk name. Would a custom chunk engine be capable of this?

I think this would be possible yet as you already tried. For interactive usage, there will be the limitation of the IDE integration I believe. Which you already found I think.

> A tarchetypes function that crawls through one or more Rmd files and translates the code chunks into _targets.R etc. for a targets pipeline. This requires the ability to purl an Rmd in a way that separately identifies the code and options of each chunk. Does rmarkdown or knitr expose a way to do this?

I believe this is possible but may use current internal function. `purl` can keep the option in the output document but currently that would mean parsing the R code for those comment. I believe this would benefit some improvmement in knitr maybe. For parsing rmd, **parsermd** is rather useful. 

> Somehow incorporate the user's original prose in rendered Rmd's that can print out the return values of targets. This last part is less clear to me, but maybe we could programmatically replace code chunks with tar_read() and render them as tar_render() outputs from the pipeline.

You could replace the code chunk content I believe. Does the `tar_read()` replacement needs to be evaluated ? or just show in the resulting output as code chunk source ? Anyway, code chunk content is in `options$code`

Sorry for the delay on this, I think you already found most of your question's answer 😓

**Reply by wlandau** [reply]
See https://community.rstudio.com/t/suppress-evaluation-of-knitr-chunk-options-in-custom-language-engine/105774/6. I am pivoting the direction of Target Markdown to https://community.rstudio.com/t/suppress-evaluation-of-knitr-chunk-options-in-custom-language-engine/105774/7?u=wlandau because it is more powerful and less likely to conflict with the features and limitations of `knitr`.

**Comment by wlandau** [comment]
A few more thoughts on this R Markdown interface idea:

## Rendered reports

What should the actual rendered report do? Since most code chunks will be individual targets, one obvious choice is to print all the targets just below their respective code chunks. This would flow nicely an naturally from a literate programming perspective, but we should limit the targets printed in case the data is too large. We could either disable reading and printing by default, or we could only print small non-dynamic targets. I like the latter.

## How to implement the rendered report

I need to read up on custom language engines and custom chunk engines. 

Naively, we could create a temporary copy of the report that runs `tar_read_raw()` in each target chunk instead of the command.

## Code chunk behavior

When run interactively in notebook mode, a chunk should run with no side effects (maybe inside `local()`) assign the return value to an object with the same name as the chunk, and print the object to the screen. These guardrails would enforce the purity and immutability requirements of `targets` and counteract the dangerous R-Markdown-oriented looseness that folks would not otherwise be wary of.

## Other files

Besides the output HTML, we know we need to generate a `_targets.R` file. Beyond that, there is an opportunity to enforce some degree of modularity. But I hesitate to do so because there is no one-size-fits-all way to organize files, the report does not provide enough structure anyway, and the unpredictable file names might disrupt projects. At the risk of creating a monolithic dumping ground for messy code (a vice which literate programming enables with impunity) I think it would be okay to go with a single `_targets.R` file, at least to start.

**Reply by wlandau** [reply]
## Guardrails

1. The generated `_targets.R` file should be treated as read-only in the RStudio IDE, e.g. https://github.com/ropensci/targets/blob/a1c5e975836c0d424fdf6274c6da4b387a2b102b/R/tar_renv.R#L98.
1. We should have some way to nudge users away from creating multiple pipeline-generating notebooks in the same working directory.

**Reply by nviets** [reply]
Thanks @wlandau ! For generating _targets.R, it might be worth looking into [knitr's purl](https://bookdown.org/yihui/rmarkdown-cookbook/purl.html), which allows the code from chunks to selectively be written out to an external file.

**Reply by wlandau** [reply]
Yeah, on reflection this feature may end up looking something like a glorified `purl`, depending on what turns out to be possible through output and language engines.

**Comment by wlandau** [comment]
Some chunk options (e.g. `pattern`) should be taken as language objects and not evaluated as R code. I can make that happen in non-interactive mode using the `eval.after` knit option and `knitr::engine_output()`. But if I run chunks in the RStudio IDE, options like `pattern` are executed like ordinary R code. Any ideas on how to suppress this? Filed an issue at https://github.com/rstudio/rstudio/issues/9407.

![Screen Shot 2021-05-25 at 10 32 21 AM](https://user-images.githubusercontent.com/1580860/119516849-19500600-bd45-11eb-9fb2-32f9443d6a71.png)

**Reply by cderv** [reply]
How does `pattern` is processed ? 

I believe chunk options in **knitr** are supposed to be evaluated. Using `eval.after` is kind of a trick as this is made just to delay evaluation (e.g fig caption using an object defined in chunk) - not really cancelled it by suppressing the option in code processing. Chunk option are just R code so, option that should evaluate to language object should be language object in R also, shouldn't it ?  

Trying to understand better how this should work at low level.

``` r
# error like in chunk
option1 <- map(data)
#> Error in map(data): impossible de trouver la fonction "map"
# object
option2 <- expression(map(data))
option2
#> expression(map(data))
is.language(option2)
#> [1] TRUE
# rlang ? 
option3 <- rlang::expr(map(data))
option3
#> map(data)
is.language(option3)
#> [1] TRUE
```

<sup>Created on 2021-05-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

what you're looking feels like a NSE chunk option ? Something like that ? 

Idea you may have tried and maybe not helpful : Does options hook could help you process the option the way you want ? 

I'll look closer into the engine you wrote to see how this is processed. 

By the way, what you encounter with the IDE is an IDE thing: `eval.after` is not working as you would expect with knitr. I believe chunk option are evaluated as code by interactive chunk mechanism as you saw in the code.

**Reply by wlandau** [reply]
The `pattern` argument drives dynamic branching: https://books.ropensci.org/targets/dynamic.html. It is a DSL that allows users to define new targets while the pipeline is running. We do not know in advance which targets or how many targets will be created because that all depends on upstream targets. In the following example, `y` is a dynamic target that depends on target `x`.

```{r}
# _targets.R file
library(targets)
list(
  tar_target(x, seq_len(round(runif(10)))),
  tar_target(y, 100 * x, pattern = map(x))
)
```

We cannot predict precisely how many `y` branches we will have, but we do know that we will end up with one branch of `y` for every vector element of `x`.

```{r}
# R console
tar_make()
#> • start target x
#> • built target x
#> • start branch y_29239c8a
#> • built branch y_29239c8a
#> • start branch y_7cc32924
#> • built branch y_7cc32924
#> • start branch y_bd602d50
#> • built branch y_bd602d50
#> • built pattern y
#> • end pipeline
```

When a target is defined,  `pattern = map(x)` is quoted and turned into an expression object.

```{r}
target_object <- tar_target(y, 100 * x, pattern = map(x))
target_object$settings$pattern
#> expression(map(x))
```

The expression `map(x)` does not actually run until the pipeline is running and it is time to process target `y` (only after processing target `x`). And when it does run, it runs only in a [special environment](https://github.com/ropensci/targets/blob/main/R/class_dynamic.R) that defines special versions of `map()`, `cross()`, etc. that act on special representations of upstream dependency targets (each dependency target is represented as a one-column data frame of branch/bud names). 

So the `pattern` argument is a DSL that needs special delayed processing.

For Target Markdown code chunks, I thought I had a workaround for interactive mode:

1. Set `pattern` in `eval.after`.
2. Take control of how `output$code` is evaluated, so evaluation behaves like a target.

https://github.com/ropensci/targets/blob/434832fbca8b525a34b810a952cca185a0c2af40/R/tar_engine_target.R#L97-L103

3. Set `eval = FALSE` for further processing so the `eval.after` chunk options like `pattern` never get evaluated.

https://github.com/ropensci/targets/blob/434832fbca8b525a34b810a952cca185a0c2af40/R/tar_engine_target.R#L109

4. Use `knitr::engine_output()` to wrap up:

https://github.com/ropensci/targets/blob/434832fbca8b525a34b810a952cca185a0c2af40/R/tar_engine_target.R#L110

But as you know from https://github.com/rstudio/rstudio/issues/9407, IDEs may ignore `eval.after` and incorrectly evaluate `pattern`.

Maybe this snag is minor: not all targets use dynamic branching, and non-interactive mode is totally unaffected. Still, other target factories may have quoted arguments, and this same issue could present other problems. With `pattern` and potential arguments like it, I have to make a decision:

1. Go forward as planned, with the understanding that `pattern` will break, with no ability to throw an informative error message (because the engine breaks). Or,
2. Encourage users to put language arguments in quotes, e.g. `{tar_target y, pattern = "map(x)"}`. Or,
3. Avoid chunk options altogether and choose a different interface for supplying target arguments.

I don't like (3) because chunk options are idiomatic and mostly convenient, despite the bugs and limitations. And I don't like (2) because the `pattern` argument of the real `tar_target()` function does not accept character strings, and this little inconsistency could confuse users. At least with (1), users will see an error in the console with interactive mode, and there may yet still be some sort of workaround. If we could at least get ahead of the chunk options processing just to prevent errors, then we could put the engines in control and be home free.

**Reply by wlandau** [reply]
> Idea you may have tried and maybe not helpful: Does options hook could help you process the option the way you want?

I tried, but it does not seem to help. I am having trouble getting option hooks to run in interactive mode. In the following report, chunk `b` dos not call `browser()` as I would have expected based on chunk `a`.

````
---
title: "test"
output: html_document
---

```{r a}
knitr::opts_hooks$set(
  opt = function(options) {
    browser()
  }
)
```

```{r b, opt = "value"}
1
```
````

**Reply by cderv** [reply]
Thanks for the detail explanation. This is pretty advanced usage of **knitr** IMO and chunk option are not currently made for delayed evaluation. I understand that using independant options name is appealing and idiomatic from **knitr**, but know that some engines are also using known knitr's option `engine.opts=` as a chunk option to pass engine specific option as a list. Seems like something that would be between (2) and (3). 

I don't see (2) as bad (but I am not a **target** user). I feel that in your package you offer a NSE interface with delayed evalutation, and now you are trying to get the same for **knitr** chunk option which currently evaluate option at some point before the engine.

However, your main issues are with interactive usage of all this, and this is a RStudio IDE limitation. Evaluation of chunks are not equivalent interactively where the IDE handles them and when knitting where R and **knitr** handle the chunks. **knitr** will define and do a lot of things during document rendering that the IDE won't do interactively when executing code chunks as R code. So unfortunately, the usage you expect is not there yet I think .

@yihui I am inviting you to this thread in case you don't know about it. I feel this topic could be of interest to you as a broad and advanced usage of **knitr**. You may have ideas or be interested by the topic.

**Reply by wlandau** [reply]
I really appreciate your comments. From your response, it sounds like I have reached genuine limitations in the available tools (as opposed to a lack of understanding on my part). And the tip about `engine.opts` is super helpful. `engine.opts` is not entirely frictionless because everything is a character string and `opts_chunk$set()` cannot set individual defaults, but it does let some options be engine specific, it avoids name conflicts, it is idiomatic to `knitr`, the intended purpose of the option fully aligns with my use case, and usage will be easy to document clearly. I would say the advantages outweigh the user-side inconveniences.

So I am thinking a target can be defined like this:

````
```{tar_target analysis, engine.opts = list(pattern = "map(data)", error = "continue")}
run_analysis(data)
```
````

and will be equivalent to this:

```{r}
# _targets.R
# ...more code
tar_target(
  name = analysis,
  command = run_analysis(data),
  pattern = map(data),
  error = "continue"
)
```

I am confident I can implement this.

**Reply by wlandau** [reply]
And a word about names:

1. Single target engine called "targets".
2. Single code chunk option, also called "targets", which controls whether we run the code as a bunch of targets or a bunch of setup globals/options.

Clean and with minimal risk of name conflicts.

````
```{targets script_name, targets = FALSE}
my_data_analysis_function <- function() {...}
```

```{targets other_script_name, targets = TRUE}
tar_target(analysis, my_data_analysis_function(data))
```
````

**Reply by wlandau** [reply]
Does `knitr` still enforce unique chunk names?

**Reply by cderv** [reply]
You mean chunk label ? 

If so, this is an option set by default to error if duplicate labels exists
https://bookdown.org/yihui/rmarkdown-cookbook/duplicate-label.html

Safety Guards because this could cause issues. And also because chunk labels are like ids and assume to be unique. For example a label name can allow to reuse some code from a chunk in another one: https://bookdown.org/yihui/rmarkdown-cookbook/reuse-chunks.html

Do you need duplicate labels for this feature ?

**Reply by wlandau** [reply]
> You mean chunk label ?

Yes, I would prefer to enforce the uniqueness of chunk labels. I would like to use these scripts as file names to write components of the `targets` pipeline in non-interactive mode. Thanks for the links, I will check them out.

**Reply by wlandau** [reply]
Looks like unique chunk guardrails are already in place by default, so we're good. Wondering if there is anything special I need to do to enforce this across multiple reports 🤔 .

**Reply by cderv** [reply]
So in this case, yes by default knitr enforce uniqueness. It will error if not. This can be manually opt-out by the user if the aforementioned option is set to `options(knitr.duplicate.label = "allow")`. So you could check this option to see if it has been modified and enforce uniques and prevent user to change this option by erroring if this is modified. I think you get the idea.

**Reply by wlandau** [reply]
Yeah, at least throwing a warning seems advisable. Thanks @cderv!

**Reply by yihui** [reply]
> I am inviting you to this thread in case you don't know about it. I feel this topic could be of interest to you as a broad and advanced usage of **knitr**. You may have ideas or be interested by the topic.

@cderv I'm afraid I don't have time to digest this discussion. It will be great if you or @wlandau could give me a brief summary (a video chat is also okay). I'm willing to provide the necessary support on **knitr**'s side.

**Reply by cderv** [reply]
I was under the impression you would say that and I did not aim for you to digest all this. Time is precious 😄 
Quick summary of the result in https://books.ropensci.org/targets/markdown.html and the source is in here : https://github.com/ropensci/targets/blob/f16cbe891b05dde8ea2a9b2f3f2e29de2e65f487/R/utils_knitr.R

but I would welcome a video chat too. @wlandau can we make that happen ? Would you be willing to a quick demo of what you end up doing ? And maybe the remaining issues or workaround needed ?

**Reply by wlandau** [reply]
Target Markdown is a `knitr`-based interface for `targets`. From `knitr`'s point of view, this is a reimagining of the  autodep and cache functionality. `targets` is well-suited for these features because of its dedicated focus on orchestration and caching. From the viewpoint of `targets`, `knitr` is a convenient interface that, when applied in the right places, does not sacrificing any of the functional programming assumptions (immutability or purity) required to detect when expensive computations can be skipped.

I would love to arrange a video chat because it is an opportunity to discuss the use case, and I sent a Doodle poll to help us align on a time. However, I think most of the basic issues are now ironed out, thanks to Christophe's help. (I was originally trying to do fancy stuff with chunk options, but workarounds are now in place that align well with `knitr`'s existing capabilities.) So I understand if a video call is no longer worth your time.

The only remaining obstacle that comes to mind is with `bookdown`. Christophe informed me that `bookdown` does not allow underscores in chunk labels. Are there plans to relax this restriction? In some Target Markdown code chunks (specifically ones with `tar_simple = TRUE`) chunk labels automatically become target names. Target names need to be valid symbol names because of how `targets` abstracts output as R objects, and the underscore is a common delimiter used in several target factories.

**Reply by yihui** [reply]
I guess it won't be too hard to remove the restriction on underscores in chunk labels. Underscores are not allowed only because `_` is a special character in LaTeX and needs to be escaped.

@cderv I guess the only place that we need to modify in bookdown is here if we want to support `_`: https://github.com/rstudio/bookdown/blob/c8883c9bbd70400144b396b7cfc66e5e4f2ae8a5/R/latex.R#L117-L130 `\@ref(foo_bar)` would become `@ref(foo\_bar)` in `.tex` output. We need to consider this `\` before `_`.

**Reply by wlandau** [reply]
Another detail: for a custom language engine, how do I print arbitrary objects (plots, `lm()` objects, etc.) instead of being limited to just text? In `knitr::engine_output()`, it looks like the `out` argument needs to be a character vector. I tried switching back to the R engine (`options$engine <- "r"`) just before calling `engine_output(out = plot_object)`, but `out` is still printed as text. I also tried manually calling `print()` on the object from inside the engine, but in the [notebook interface](https://bookdown.org/yihui/rmarkdown/notebook.html), some outputs get printed to the R console instead of inline in the notebook.

**Reply by wlandau** [reply]
And thanks so much @yihui for your willingness to allow underscores in `bookdown` chunk labels!

**Reply by yihui** [reply]
> Another detail: for a custom language engine, how do I print arbitrary objects (plots, `lm()` objects, etc.) instead of being limited to just text?

The `out` argument can be a list instead of a character vector. The key is that you should not provide the `code` argument to `engine_output()`. In that case, **knitr** calls the S3 generic function `knitr::sew()` to generate a character vector (output): https://github.com/yihui/knitr/blob/master/R/engine.R#L82 You can register custom S3 methods.

I'm not sure if that answers your question. Perhaps I can help better if you show me an example on Monday.

**Reply by wlandau** [reply]
I just tried to implement your suggestions in https://github.com/ropensci/targets/commit/a77373182961b8ec83ddb55549607b156017c001: i.e.
when `tar_simple = TRUE`, invoke `knitr::engine_output()` with `out` as a list and with no value for `code`.

https://github.com/ropensci/targets/blob/a77373182961b8ec83ddb55549607b156017c001/R/utils_knitr.R#L124-L128

I tested it with this minimal report:

````
---
title: "Target Markdown"
output: html_document
---

```{r setup, include = FALSE}
library(targets)
tar_option_set(packages = "ggplot2")
```

```{targets targetname, tar_simple = TRUE}
ggplot(mtcars) + geom_histogram(aes(x = mpg))
```
````

I opened this report in the RStudio IDE and pressed the play button for each chunk sequentially. I expected the `{targets targetname}` chunk to print a `ggplot` object inline just like the R engine. But instead, the message "don't know how to handle 'targets' engine output" printed to the R console. I find this puzzling because `options$engine` gets changed to `"r"` just before `knitr::engine_output(options = options, out = out)`.

This is not the most crucial feature, and admittedly I have some mixed feeling about implementing it. On the other hand, the behavior is not what I expected from `knitr::engine_output()`, especially because the R engine seems to print all kinds of values inline in the notebook and respect the `print()` method of the particular object.

**Reply by wlandau** [reply]
~~Maybe I should just try an output hook?~~ On the other hand, the default `output` and `plot` hooks already look right, so maybe not.

**Comment by wlandau** [comment]
I love how this feature turned out! Resources:

* https://books.ropensci.org/targets/markdown.html
* Target Markdown template: https://rstudio.github.io/rstudio-extensions/rmarkdown_templates.html
* A new `use_targets()` function.

My last question is about syntax highlighting. Is there a way to use the same syntax highlighting for the {targets} engine as the {r} engine?

**Reply by cderv** [reply]
In the RStudio IDE, this is a matter of IDE support. I am not sure how they detect the language to highlight in the chunk but I suspect the chunk engine.

Regarding the HTML output for the source code, it is derived from the engine name. So his is a matter of how your engine outputs. I need to look at your code to know specifically how in your case. But basically, you may be able to modify the correct option after you engine has done its work. 

Or we need to teach knitr about your engine name 😉

**Reply by cderv** [reply]
So I believe your engine is this one 
https://github.com/ropensci/targets/blob/82ead80b0250e6691e9a623234b5fa92f3147f1c/R/tar_engine.R#L68-L71

So you could do
```diff
tar_engine_output <- function(options, out) { 
  code <- paste(options$code, collapse = "\n") 
+ # to get the correct markup later in knitr source hook
+ options$engine <- 'r'
  knitr::engine_output(options = options, code = code, out = out) 
} 
```
This after your engine has done its work, so I don't think changing the engine will have impact on how the code is evaluated. It will just be used for further processing, and which class to apply on fenced code attribute in Markdown output is one of them.

I am a bit lazy as it is Friday night here in France so I did not try :) I let you try ? 

Other solution would be for you to set the `class.source` option to `r`, through your engine or using an option hook (but it seems the `targets` chunk option is not to be provided each time)
https://bookdown.org/yihui/rmarkdown-cookbook/chunk-styling.html#chunk-styling

Adding the class is what triggers the Pandoc highlighting (https://pandoc.org/MANUAL.html#fenced-code-blocks)

If this is not working as I thought, we'll surely be able to do something (maybe adding a `language` options or similar.

**Reply by wlandau** [reply]
Thanks so much, Christophe! `options$engine <- "r"` is perfect! Works out of the box.
